### PR TITLE
[ASR] add cache-aware streaming transformer

### PR DESCRIPTION
**Pull request by stevehuang52**
# Add StreamingTransformerEncoder: cache-aware streaming Transformer encoder for ASR

## What does this PR do?

Adds `StreamingTransformerEncoder`, a convolution-free Transformer encoder for streaming ASR. It
extends the existing `TransformerEncoder` with:

1. **Bounded local attention** via `att_context_size: [left, right]` (encoder frames) under two
   styles:
   - `chunked_limited` (default) — chunk-aligned attention matching `ConformerEncoder._create_masks`
     bit-for-bit. Look-ahead does **not** compound across layers, so streaming stays exact at any
     depth.
   - `sliding_window` — `[t - left, t + right]`; exact for streaming only when
     `right == 0`, since a sliding right context compounds across layers.
2. **Cache-aware streaming** with a rolling KV cache, implementing the real `StreamingEncoder`
   interface (`cache_aware_stream_step` / `get_initial_cache_state` / `setup_streaming_params`)
   and reusing the Conformer `cache_last_channel` plumbing, so the shared streaming tooling drives
   it unchanged.
3. **Multi-latency training** — a list of `[left, right]` pairs samples one context per training
   batch (mirroring `ConformerEncoder`); eval and streaming use a pinned context.

The cache stores raw, position-agnostic per-layer inputs, so the format is independent of the
positional scheme. Streaming supports `rel_pos`, `rope` and `no_pos` (`abs_pos` raises).

## Changes

| file | change |
|---|---|
| `nemo/collections/asr/modules/transformer_encoder.py` | `StreamingTransformerEncoder`, the chunked/sliding mask mods, and the streaming step |
| `nemo/collections/asr/modules/__init__.py` | export `StreamingTransformerEncoder` |
| `nemo/collections/asr/parts/submodules/subsampling.py` | `FeatureStacking.get_sampling_frames` / `get_streaming_cache_size`, probed by the cache-aware streaming utils |
| `examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml` | RNN-T training recipe |
| `docs/source/asr/api.rst` | autoclass entries for `TransformerEncoder` and `StreamingTransformerEncoder` |
| `docs/source/asr/configs.rst` | "Streaming Transformer" section covering `att_context_size` / `att_context_style` / `pre_encode_cache_size` |
| `tests/collections/asr/test_transformer_encoder.py` | 124 unit tests |

No existing behaviour changes: `TransformerEncoder` is untouched apart from an overridable
`_build_mask_mod` hook, and the `subsampling.py` additions are new methods only.

## Evaluation

Trained a 0.6 B RNN-T with this encoder (48 layers, `d_model` 1024, `feature_stacking` ×8, `rope`,
`chunked_limited`, contexts `[[70,13],[70,6],[70,1],[70,0]]`) and compared against the released
`nvidia/nemotron-speech-streaming-en-0.6b` FastConformer RNN-T. Preprocessor, SpecAugment, RNN-T
decoder, joint, loss, vocab, context ladder and parameter count are matched; the encoder is the
variable. Training data differs and is uncontrolled.

8 Open ASR Leaderboard test sets (82,185 utts / 167.6 h), byte-identical eval sets for both
models, scored with the HuggingFace Open ASR Leaderboard normalizer. Harness validated: the
reference reproduces its published model card (ls-other 5.18 vs 4.84, spgispeech 3.11 vs 2.97,
gigaspeech 8.93 vs 9.66).

**Offline inference with `examples/asr/transcribe_speech.py` (macro WER%, lower is better)**

| chunk | lookahead | this PR | FastConformer |
|---|---|---|---|
| 14 | 1.12 s | **6.78** | 6.85 |
| 7 | 0.56 s | **6.94** | 6.97 |
| 2 | 0.16 s | 7.70 | **7.43** |

At parity — this encoder wins 5 of 8 datasets at chunks 14 and 7.

**True cache-aware streaming using `examples/asr/asr_streaming_inference/asr_streaming_infer.py` (macro WER penalty vs offline)**

| chunk | this PR | FastConformer |
|---|---|---|
| 14 | **+0.01** | +0.40 |
| 7 | **+0.02** | +0.37 |
| 2 | **+0.12** | +0.51 |

Streaming is essentially free for this encoder. At chunk 2 this flips the ranking: offline the
Conformer leads 7.43 vs 7.70, but under real streaming this PR leads **7.82 vs 7.94**. Hypothesis
agreement with the offline forward is 86–93% here vs 71–80% for the Conformer.

## Testing

`pytest tests/collections/asr/test_transformer_encoder.py` — **124 passed**.

Coverage includes:
- `chunked_limited` mask parity with `ConformerEncoder._create_masks`, elementwise, for every
  configured context and several sequence lengths
- `test_streaming_matches_full_forward` — chunk-by-chunk streaming equals the full forward
  (~1e-6 fp32), including ragged batches, partial final chunks, cache warm-up and rolling caches
- multi-context sampling (train-only, deterministic in eval), cache shapes, `abs_pos` guard,
  `bypass_pre_encode`, activation-checkpointing unwrap, dropout wiring

Docs build: new `autoclass` targets and the `:ref:` labels they define resolve.

### CODE REVIEW

File: examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    joint_hidden: 640
+
+  train_ds:
+    manifest_filepath: ???
+    sample_rate: ${model.sample_rate}
+    batch_size: 16 # you may increase batch_size if your memory allows
+    shuffle: true
+    num_workers: 8
+    pin_memory: true
+    max_duration: 20 # you may need to update it for your dataset
```
Comment (nithinraok) [inline examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml]: lets change to 40 by default?

File: examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    type: bpe  # Can be either bpe (SentencePiece tokenizer) or wpe (WordPiece tokenizer)
+
+  preprocessor:
+    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
+    sample_rate: ${model.sample_rate}
+    normalize: "NA" # No normalization for mel-spectogram makes streaming easier
+    window_size: 0.025
+    window_stride: 0.01
+    window: "hann"
+    features: 80
```
Comment (nithinraok) [inline examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml]: update to 128. Have you used 80 or 128 for your experiments

File: docs/source/asr/configs.rst [hunk 3]
```diff
@@ -432,6 +432,53 @@ Conformer-Transducer
...
+      subsampling_factor: 8
+      self_attention_model: rope     # rel_pos | abs_pos | rope | no_pos
+      qk_norm: true
+      att_context_style: chunked_limited
+      att_context_size: [[70, 13], [70, 6], [70, 1], [70, 0]]
+      # Input-frame look-back prepended to each streaming chunk so the mel front-end's STFT
+      # window has its context. Must be a multiple of ``subsampling_factor``. ``FeatureStacking``
+      # is non-overlapping and needs none of this offline, but a streaming front-end that
+      # recomputes the spectrogram per chunk does.
+      pre_encode_cache_size: ${model.encoder.subsampling_factor}
```
Comment (nithinraok) [inline docs/source/asr/configs.rst:484]: can we also add info about cache_size

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 4]
```diff
@@ -121,20 +125,79 @@ def causal(b, h, q_idx, kv_idx):
...
+        # No dropout after the output projection: ``TransformerBlock`` already applies
+        # dropout to this module's output on the residual branch, and stacking the two
+        # gives an effective rate of ``1 - (1 - drop_rate)^2`` (19% at the default 0.1)
+        # with ~2.1x the intended noise variance, in every layer. Matches
+        # ``ConformerFeedForward``, which also drops only after the activation.
+        #
+        # Keep the trailing position free rather than renumbering: the module indices are
+        # state_dict keys (``net.0.*``, ``net.3.*``), so removing this last entry stays
+        # load-compatible with existing checkpoints, while dropping the *inner* Dropout
+        # would shift the second Linear to ``net.2.*`` and break them.
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder.py]: good find. Lets remove this comment.

Comment (nithinraok) [reply nemo/collections/asr/modules/transformer_encoder.py]: or write a one liner.

File: examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+      score_norm: true
+      tsd_max_sym_exp: 50  # for Time Synchronous Decoding
+      alsd_max_target_len: 2.0  # for Alignment-Length Synchronous Decoding
+
+  # config for InterCTC loss: https://arxiv.org/abs/2102.03216
+  # specify loss weights and which layers to use for InterCTC
+  # e.g., to reproduce the paper results, set loss_weights: [0.3]
+  # and apply_at_layers: [8] (assuming 18 layers). Note that final
+  # layer loss coefficient is automatically adjusted (to 0.7 in above example)
+  interctc:
```
Comment (nithinraok) [inline examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml]: we don;t have support for this yet?

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 6]
```diff
@@ -624,3 +834,514 @@ def freeze(self) -> None:
...
+            raise ValueError(
+                f"Overlapping/striding chunks are not supported: shift_size ({shift}) must equal "
+                f"chunk_size ({chunk}). Every output frame of this encoder already equals the "
+                f"offline forward, so there is nothing for an overlap to recompute."
+            )
+
+        if left_chunks is not None:
+            # Retune the window itself, not just the cache — otherwise the mask and the rolling
+            # cache would describe different amounts of history.
+            self.att_context_size[0] = int(left_chunks) * chunk
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder.py:1161]: this will overwrite left context?

File: examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml (hunk 2 above)
Comment (stevehuang52) [reply examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml]: I used 128, this one was just copied from earlier FC configs by swapping the encoder

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 6 above)
Comment (stevehuang52) [reply nemo/collections/asr/modules/transformer_encoder.py:1161]: yeah it's used to override the left context size during inference, but in practice we always use the left context seen in training so this `left_chunks` is always `None` by default

File: examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml (hunk 5 above)
Comment (stevehuang52) [reply examples/asr/conf/transformer/transformer_stacking_rnnt_bpe_streaming.yaml]: I think we need to switch to hybrid-ctc-rnnt model, otherwise this section is no-op I think

File: docs/source/asr/configs.rst (hunk 3 above)
Comment (stevehuang52) [reply docs/source/asr/configs.rst:484]: The more detailed docstring is in the `StreamingTransformerEncoder` class, overall we just need it to be equal to subsampling_factor, setting it to 0 has about 0.1 WER degradation.

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 7]
```diff
@@ -272,6 +325,108 @@ def forward(self, x, block_mask=None, pos_emb=None):
...
+        if self.self_attention_model == "no_pos":
+            return q, k, None
+        # abs_pos bakes absolute positions into the embeddings before the blocks, which does
+        # not compose with a running KV cache without a position offset; deferred.
+        raise NotImplementedError(
+            f"Cache-aware streaming is not implemented for self_attention_model="
+            f"'{self.self_attention_model}'. Supported streaming schemes: 'rel_pos', 'rope', 'no_pos'."
+        )
+
+    def _build_streaming_rel_pos_score_mod(self, q, pos_emb, num_cur, num_kv):
```
Comment (tango4j) [inline nemo/collections/asr/modules/transformer_encoder.py:385]: Missing docstring like in the docstring in `forward_streaming`:

Please add docstrings with the following format - 

Args:
     var_name1 (type): descriptions
     var_name2 (type): descriptions

Returns: 
     var_name1 (type): descriptions
     var_name2 (type): descriptions

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 8]
```diff
@@ -272,6 +325,108 @@ def forward(self, x, block_mask=None, pos_emb=None):
         out = out.transpose(1, 2).contiguous().view(B, T, self.d_model)
         return self.out_proj(out)
 
+    def _apply_streaming_position(self, q, k, pos_emb, num_cur, num_kv):
```
Comment (tango4j) [inline nemo/collections/asr/modules/transformer_encoder.py:342]: Missing docstring like in the docstring in forward_streaming:

Please add docstrings with the following format -

Args:
var_name1 (type): descriptions
var_name2 (type): descriptions

Returns:
var_name1 (type): descriptions
var_name2 (type): descriptions

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 9]
```diff
@@ -121,8 +125,58 @@ def causal(b, h, q_idx, kv_idx):
     return causal
 
 
+def _make_sliding_window_mod(left, right):
```
Comment (tango4j) [inline nemo/collections/asr/modules/transformer_encoder.py:128]: Missing docstring like in the docstring in `forward_streaming()` function:

Please add docstrings with the following format -

Args:
var_name1 (type): descriptions
var_name2 (type): descriptions

Returns:
var_name1 (type): descriptions
var_name2 (type): descriptions

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 10]
```diff
@@ -121,8 +125,58 @@ def causal(b, h, q_idx, kv_idx):
...
+        if left >= 0 and right >= 0:
+            return (kv_idx >= q_idx - left) & (kv_idx <= q_idx + right)
+        if left >= 0:
+            return kv_idx >= q_idx - left
+        return kv_idx <= q_idx + right
+
+    return sliding_window
+
+
+def _make_chunked_limited_mod(left, right):
```
Comment (tango4j) [inline nemo/collections/asr/modules/transformer_encoder.py:157]: Missing docstring like in the docstring in `forward_streaming()` function:

Please add docstrings with the following format -

Args:
var_name1 (type): descriptions
var_name2 (type): descriptions

Returns:
var_name1 (type): descriptions
var_name2 (type): descriptions

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 11]
```diff
@@ -594,6 +779,21 @@ def forward_internal(self, audio_signal, length, bypass_pre_encode=False):
         length = length.to(dtype=torch.int64)
         return x, length
 
+    def _build_mask_mod(self, length):
```
Comment (tango4j) [inline nemo/collections/asr/modules/transformer_encoder.py:817]: Missing docstring like in the docstring in `forward_streaming()` function:

Please add docstrings with the following format -

Args:
var_name1 (type): descriptions
var_name2 (type): descriptions

Returns:
var_name1 (type): descriptions
var_name2 (type): descriptions

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 12]
```diff
@@ -624,3 +824,514 @@ def freeze(self) -> None:
...
+        """
+        if att_context_size is not None:
+            if len(att_context_size) != 2:
+                raise ValueError(f"att_context_size must be a [left, right] pair, but got {att_context_size}.")
+            self.att_context_size = list(att_context_size)
+
+        # Encoder frames per step. ``right + 1`` is the chunk under chunked_limited and the
+        # minimum-latency step under a causal sliding window (right == 0 -> 1 frame).
+        implied_chunk = self.att_context_size[1] + 1 if self.att_context_size[1] >= 0 else 1
+        chunk = implied_chunk if chunk_size is None else int(chunk_size)
```
Comment (tango4j) [inline nemo/collections/asr/modules/transformer_encoder.py:1140]: This part seems to need one more error-checking line:

if self.att_context_style == "chunked_limited" and chunk > implied_chunk:

The evidence is:

COunter example 1 )


att_context_size = [8, 3]
chunk_size = 2
shift_size = 2

Then the code becomes 

right = 3
implied_chunk = right + 1 = 4
chunk > implied_chunk
2 > 4 ---> False

=============
Could you re-check if this additional check is needed ? 
If you are limiting this value somewhere else, it is not an issue.

**Review by tango4j — CHANGES_REQUESTED** [review]
There are some docstrings missing in the functions. 
And one part to check the logic.

**Review by nithinraok — APPROVED** [review]
LGTM

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 12 above)
Comment (stevehuang52) [reply nemo/collections/asr/modules/transformer_encoder.py:1140]: Updated.

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 11 above)
Comment (stevehuang52) [reply nemo/collections/asr/modules/transformer_encoder.py:817]: added all missing docstrings

**Review by tango4j — APPROVED** [review]
All the requested changes are reflected. Approving
